textdb/textdb-dataflow/src/main/java/edu/uci/ics/textdb/dataflow/regexmatch/RegexMatcher.java

```diff
@@ -34,23 +51,46 @@ public void open() throws DataFlowException {
 
     @Override
     public ITuple getNextTuple() throws DataFlowException {
-
-        try {
+		try {
             ITuple sourceTuple = sourceOperator.getNextTuple();
             if(sourceTuple == null){
                 return null;
             }
-            if (predicate.satisfy(sourceTuple)) {
-                return sourceTuple;
-            } else {
-                return getNextTuple();
+            
+            RegexPredicate rPredicate = (RegexPredicate)predicate; 
+            
+            spans = rPredicate.statisfySpan(sourceTuple);
+            
+            if (spans.size() != 0) { // a list of matches found
+            	schema = sourceTuple.getSchema();
+            	fields = sourceTuple.getFields();
+            	return getSpanTuple(spans);
+            } else { // no match found
+            	return getNextTuple();
             }
         } catch (Exception e) {
             e.printStackTrace();
             throw new DataFlowException(e.getMessage(), e);
-        }
-
+        }        
+    }
+    
+    private Schema createSpanSchema() {
+    	List<Attribute> attributesCopy = new ArrayList<>(schema.getAttributes());
+    	attributesCopy.add(SchemaConstants.SPAN_LIST_ATTRIBUTE);
+    	return new Schema(attributesCopy.toArray(new Attribute[attributesCopy.size()]));
     }
+    
+    private ITuple getSpanTuple(List<Span> spans) {
+    	Schema spanSchema = createSpanSchema();
```

Comment:
This "createSpanSchema()" is called each time "getSpanTuple()" is called.  That's too expensive.  Can you create it only once?